sql/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/LocalRelationDatasource.scala

```diff
@@ -0,0 +1,152 @@
+import org.apache.spark.sql.vectorized.{ArrowColumnVector, ColumnarBatch, ColumnVector}
+import org.apache.spark.storage.CacheId
+
+/**
+ * Datasource implementation that can read a Connect Chunked Datasource directly from the
+ * block manager.
+ *
+ * @param id of the LocalRelation.
+ * @param sessionId of the LocalRelation.
+ * @param schema of the LocalRelation.
+ * @param dataHashes ids of the blocks stored in the BlockManager.
+ */
+class LocalRelationTable(
+    id: Long,
+    sessionId: String,
+    schema: StructType,
+    dataHashes: Seq[String])
+  extends Table with SupportsRead {
+
+  override def name(): String = s"LocalRelation[session=$sessionId, id=$id]"
+
+  override def columns(): Array[Column] = schema.fields.map { field =>
+    Column.create(field.name, field.dataType, field.nullable)
+  }
+
+  override def capabilities(): util.Set[TableCapability] = util.EnumSet.of(
+    TableCapability.BATCH_READ,
+    TableCapability.MICRO_BATCH_READ)
+
+  override def newScanBuilder(options: CaseInsensitiveStringMap): ScanBuilder = new ScanBuilder {
+    override def build(): Scan = new LocalRelationScan(schema, sessionId, dataHashes)
+  }
+}
+
+case class LocalRelationScanPartition(sessionId: String, dataHash: String) extends InputPartition
+
+class LocalRelationScan(schema: StructType, sessionId: String, dataHashes: Seq[String])
+  extends Scan with Batch {
+  override def readSchema(): StructType = schema
+
+  override def toBatch: Batch = this
+
+  override def planInputPartitions(): Array[InputPartition] =
+    dataHashes.map(hash => LocalRelationScanPartition(sessionId, hash)).toArray
+
+  override def createReaderFactory(): PartitionReaderFactory =
+    LocalRelationScanPartitionReaderFactory
+
+  override def columnarSupportMode(): Scan.ColumnarSupportMode =
+    Scan.ColumnarSupportMode.SUPPORTED
+}
+
+object LocalRelationScanPartitionReaderFactory extends PartitionReaderFactory {
+  override def supportColumnarReads(partition: InputPartition): Boolean = true
+
+  override def createReader(partition: InputPartition): PartitionReader[InternalRow] =
+    throw new IllegalStateException("Row based reads are not supported (yet)")
+
+  override def createColumnarReader(partition: InputPartition): PartitionReader[ColumnarBatch] = {
+    val localRelationScanPartition = partition.asInstanceOf[LocalRelationScanPartition]
+    new LocalRelationPartitionReader(localRelationScanPartition)
+  }
+}
+
+class LocalRelationPartitionReader(partition: LocalRelationScanPartition)
+  extends PartitionReader[ColumnarBatch] with Logging {
+  private var input: InputStream = _
+  private var allocator: BufferAllocator = _
+  private var reader: ConcatenatingArrowStreamReader = _
+
+  private def init(): Unit = {
+    if (reader == null) {
+      val blockManager = SparkEnv.get.blockManager
+      val blockId = CacheId(partition.sessionId, partition.dataHash)
+      input = blockManager.getLocalBytes(blockId).map(_.toInputStream())
```

Comment:
I think this can be improved. I don't think we replicate the block to this executor by doing the remote read like this. If we were to replicate, then we could also try to use locality to make re-reads more effective.